cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -711,6 +718,217 @@ class self_comparator {
+    {
+    }
+
+    template <typename T, CUDF_ENABLE_IF(not cudf::is_nested<T>())>
+    __device__ hash_value_type operator()(column_device_view const& col,
+                                          size_type row_index) const noexcept
+    {
+      return element_hasher<hash_fn, Nullate>(_has_nulls, _seed, NULL_HASH)
+        .template operator()<T>(col, row_index);
+    }
+
+    template <typename T, CUDF_ENABLE_IF(cudf::is_nested<T>())>
+    __device__ hash_value_type operator()(column_device_view const& col,
+                                          size_type row_index) const noexcept
+    {
+      auto hash                   = hash_value_type{_seed};
+      column_device_view curr_col = col.slice(row_index, 1);
+      while (is_nested(curr_col.type())) {
+        if (_has_nulls) {
+          auto validity_it = detail::make_validity_iterator<true>(curr_col);
+          hash             = detail::accumulate(
+            validity_it, validity_it + curr_col.size(), hash, [](auto hash, auto is_valid) {
+              return cudf::detail::hash_combine(hash, is_valid ? 0 : NULL_HASH);
+            });
+        }
+        if (curr_col.type().id() == type_id::STRUCT) {
+          if (curr_col.num_child_columns() == 0) { return hash; }
+          // Non-empty structs are assumed to be decomposed and contain only one child
+          curr_col = detail::structs_column_device_view(curr_col).sliced_child(0);
+        } else if (curr_col.type().id() == type_id::LIST) {
+          auto list_col   = detail::lists_column_device_view(curr_col);
+          auto list_sizes = make_list_size_iterator(list_col);
+          hash            = detail::accumulate(
+            list_sizes, list_sizes + list_col.size(), hash, [](auto hash, auto size) {
+              return cudf::detail::hash_combine(hash, hash_fn<decltype(size)>{}(size));
+            });
+          curr_col = list_col.sliced_child();
+        }
+      }
+      for (int i = 0; i < curr_col.size(); ++i) {
+        hash = cudf::detail::hash_combine(
+          hash,
+          type_dispatcher<dispatch_void_if_nested>(
+            curr_col.type(),
+            element_hasher<hash_fn, Nullate>{_has_nulls, DEFAULT_HASH_SEED, NULL_HASH},
+            curr_col,
+            i));
+      }
+      return hash;
+    }
+
+    uint32_t _seed;
+    Nullate _has_nulls;
+  };
+
+  CUDF_HOST_DEVICE device_row_hasher(Nullate has_nulls,
+                                     table_device_view t,
+                                     uint32_t seed = DEFAULT_HASH_SEED) noexcept
+    : _table{t}, _seed(seed), _has_nulls{has_nulls}
+  {
+  }
+
+  table_device_view _table;
+  Nullate _has_nulls;
+  uint32_t _seed;
+};
+
+// Inject row::equality::preprocessed_table into the row::hash namespace
+// As a result, row::equality::preprocessed_table and row::hash::preprocessed table are the same
+// type and are interchangeable.
+using preprocessed_table = row::equality::preprocessed_table;
+
+class row_hasher {
```

Comment:
It seems like the `row_hasher` and the `self_comparator` could be implemented as template specializations of the same class, where the `device_hasher` and `device_comparator` methods would just be implemented as `device_functor<typename Functor>` or something. You'd need to find an appropriate parent namespace to store it in, but it would be nice to remove that redundancy.

Reply:
`device_hasher` and `device_comparator` differ in the logic such that the latter has a lot of early exits where the former is just meant to hash combine the result of a statement. It would be non-trivial to merge them. Such merging would be difficult to maintain in case any behaviour diverges in the future.

For the owning `row_hasher` and `self_comparator` objects, I agree the ctor code is same but they're tiny. Most of the space is taken up by docs.

Reply:
Actually sharing the docs is also part of what I had in mind here. Since we have more outstanding work coming in this direction, though, I'm inclined to wait until we have more comparators and then refactor since I suspect some of the points for refactoring will become more obvious as we add the two table comparators etc.